File: src/main/java/dev/esophose/discordbot/command/DiscordCommandArgumentInfo.java
package dev.esophose.discordbot.command;

import java.lang.reflect.Parameter;
import java.lang.reflect.ParameterizedType;
import java.util.Optional;

public class DiscordCommandArgumentInfo {

    private final Parameter parameter;

    public DiscordCommandArgumentInfo(Parameter parameter) {
        this.parameter = parameter;
    }

    public Class<?> getType() {
        return this.isOptional() ? this.getOptionalType() : this.parameter.getType();
    }

    public String getName() {
        return this.parameter.getName();
    }

    public boolean isOptional() {
        return this.parameter.getType() == Optional.class;
    }

    public Class<?> getOptionalType() {
        return (Class<?>) ((ParameterizedType) this.parameter.getParameterizedType()).getActualTypeArguments()[0];
    }

    public boolean isEnum() {
        return this.parameter.getType().isEnum();
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/DiscordCommandArgumentInfo.kt
package dev.esophose.discordbot.command

import java.lang.reflect.Parameter
import java.lang.reflect.ParameterizedType
import java.util.Optional

class DiscordCommandArgumentInfo(private val parameter: Parameter, val position: Int) {

    val type: Class<*>
        get() = if (this.isOptional) this.optionalType!! else this.parameter.type

    val name: String
        get() = this.parameter.name!!

    val isOptional: Boolean
        get() = this.parameter.type == Optional::class.java

    val optionalType: Class<*>?
        get() = (this.parameter.parameterizedType as ParameterizedType).actualTypeArguments[0] as? Class<*>

    val isEnum: Boolean
        get() = this.parameter.type.isEnum

}
